### DISCUSSION TITLE
Receiving Observations via MQTT, when those are written via HTTP

### DISCUSSION THREAD

**Discussion by mikedolx** [FraunhoferIOSB/FROST-Server#1417] [Q&A]
Hi,

i am not receiving any updates via MQTT, when large amount of Observations are imported into FROST via HTTP/REST. Is this intended behaviour? 
On the other hand, when i create an Observation with my python script (using the Python FROST client), from my dev machine i do get a published message in my MQTT-Explorer.
In both cases i communicate with the same server.

Here's the python script:
```python
import frost_sta_client as fsc
from datetime import datetime, tzinfo
import random

url = "https://mydomain.com:443/FROST-Server/v1.1"

authentication = fsc.AuthHandler("anyuser", "secretpassword")
service = fsc.SensorThingsService(url, auth_handler=authentication)

featureOfInterest = service.features_of_interest().query().list().get(0)

datastream = None
try:
    datastreamList = service.datastreams().query().filter("name eq 'Dummy-Datastream'").list()
    datastream = datastreamList.get(0)

    for _ in range(100):
        observationValue = random.randrange(1, 100)

        currentTime = datetime.now().strftime('%Y-%m-%dT%H:%M:%S.000Z')
        print(currentTime)
        observation = fsc.Observation(datastream=datastream, result_time=currentTime, phenomenon_time=currentTime, result=observationValue, 
            feature_of_interest=featureOfInterest)
        service.create(observation)

        print("creating observation of type humidity with value = " + str(observationValue) + " and id = " + str(observation.id))
except Exception as ex:
    print("Exception = " + str(ex))   
```

Do understand the concept correctly: I should receive newly published entities via MQTT independently of how the entity has been created in FROST?

**Comment by hylkevds** [comment]
You should always be receiving MQTT messages about new Observations. Even if there are many, and regardless of how the entity was created (Directly, via deep-insert, via batch-requests or via DataArrays).

When there flow is really large some messages may be lost, but FROST should complain about queue overruns when that happens, and indicate which settings control the relevant queue. Usually this only happens when using DataArrays to create Observations.

**Comment by mikedolx** [comment]
> Usually this only happens when using DataArrays to create Observations.

Do you mean Observations sent as an array to frost? Like: `{ "value": [ ... ] }`  Or is that a specific FROST-Term?

As far as i can see it from our script code, we do send multiple observations as an array to FROST. But i do not see any errors in the docker frost logs. Also i don't get any result when greping the container internal logs for "overrung":
```
/usr/local/tomcat/logs$ grep overrun *.log
```

**Comment by hylkevds** [comment]
See the DataArray extension in the specification: https://docs.opengeospatial.org/is/18-088/18-088.html#create-observation-dataarray

The message you would be looking for is: _Failed to add message to send-queue_
https://github.com/FraunhoferIOSB/FROST-Server/blob/v2.x/FROST-Server.Core/src/main/java/de/fraunhofer/iosb/ilt/frostserver/messagebus/MqttMessageBus.java#L272

But even if there are buffer overruns, you should at least get some MQTT messages. If you are getting no messages at all, they you are either using the wrong subscription, or there is another issue somewhere.

**Reply by mikedolx** [reply]
If found this snippet in our script. So i gues, we are using DataArrays:

```
 $observation_datastream = array(
                    "Datastream" => array("@iot.id" => $frost_datastream['@iot.id']),
                    "components" => array("result", "phenomenonTime", "resultTime"),
                    "dataArray" => array() );

```

The logs do not contain the key word "send-queue". 

I'm subscribing to v1.1/Observations. Again, when i create an observation with my python script, then i get this Observation published to my MQTT-Explorer. So i guess it somehow has to do with the script using DataArrays, allthough i don't see the FROST-Server complaining about it.

**Reply by hylkevds** [reply]
The default queue size is 1000, so unless you're creating more than that per POST, there should not be any errors.

Are the Observations actually created in the database?
And are you getting back a list of selfLinks for those Observations?

**Reply by mikedolx** [reply]
There should be around 1500 Observations created by a single POST. I'lll increase the queue size to 2000 for now. I will have to test it next week. Currently it seems i have no access to the FROST-Server via MQTT.

**Reply by mikedolx** [reply]
As it seems, the MQTT broker is working. We might have an issue with the batch import of our data. Ocasionally, sometimes we get a published message from a sensor, that is not imported via batch script.
Therefor, i assume MQTT is working. Nevertheless, thanks for helping to debug,
